From: kkuo <kat.kuo@canonical.com>
Date: Mon, 5 Jan 2026 16:31:08 -0500
Subject: telemetry: Repurpose diagnostics panel for all telemetry
 applications

This patch renames and repurposes the diagnostics panel to be a catchall
for telemetry-related features.

The Diagnostics panel is renamed to Telemetry while maintaining the same
ABRT and Whoopsie functionality and modularization. To reduce the patch
size, the Diagnostics panel code is deactivated, not removed.

Additional modular integration with Ubuntu Insights via the insights
build argument is also added. This integration allows for user consent
to be managed in a centralized and fine-grained manner. Additionally,
functionality to generate and show an example report is also present.
---
 meson.build                                        |   9 +
 meson_options.txt                                  |   1 +
 panels/privacy/cc-privacy-panel.blp                |  12 +-
 panels/privacy/cc-privacy-panel.c                  |   4 +-
 panels/privacy/gnome-privacy-panel.desktop.in      |   2 +-
 panels/privacy/meson.build                         |  27 +-
 ...vacy.gresource.xml => privacy.gresource.xml.in} |   4 +-
 .../telemetry/cc-telemetry-insights-example.blp    |  47 ++
 .../telemetry/cc-telemetry-insights-example.c      | 102 ++++
 .../telemetry/cc-telemetry-insights-example.h}     |  14 +-
 panels/privacy/telemetry/cc-telemetry-page.blp     |  90 +++
 panels/privacy/telemetry/cc-telemetry-page.c       | 629 +++++++++++++++++++++
 .../cc-telemetry-page.h}                           |   9 +-
 .../privacy/{diagnostics => telemetry}/meson.build |   6 +-
 panels/privacy/telemetry/telemetry.css             |   8 +
 po/POTFILES.in                                     |   4 +
 16 files changed, 944 insertions(+), 24 deletions(-)
 copy panels/privacy/{privacy.gresource.xml => privacy.gresource.xml.in} (89%)
 create mode 100644 panels/privacy/telemetry/cc-telemetry-insights-example.blp
 create mode 100644 panels/privacy/telemetry/cc-telemetry-insights-example.c
 copy panels/{mouse/cc-mouse-test.h => privacy/telemetry/cc-telemetry-insights-example.h} (62%)
 create mode 100644 panels/privacy/telemetry/cc-telemetry-page.blp
 create mode 100644 panels/privacy/telemetry/cc-telemetry-page.c
 copy panels/privacy/{camera/cc-camera-page.h => telemetry/cc-telemetry-page.h} (73%)
 copy panels/privacy/{diagnostics => telemetry}/meson.build (73%)
 create mode 100644 panels/privacy/telemetry/telemetry.css

diff --git a/meson.build b/meson.build
index 79cd508..30888b4 100644
--- a/meson.build
+++ b/meson.build
@@ -353,6 +353,14 @@ endif
 config_h.set('HAVE_WHOOPSIE', enable_whoopsie,
              description: 'Define to 1 if Whoopsie is enabled')
 
+# Insights (Ubuntu telemetry service)
+enable_insights = get_option('insights')
+if enable_insights
+  insights_dep = dependency('insights', version: '>= 0.8.1')
+endif
+config_h.set('HAVE_INSIGHTS', enable_insights,
+             description: 'Define to 1 if Ubuntu Insights is enabled')
+
 if host_is_linux
   # ModemManager
   mm_dep =  dependency('mm-glib', version: '>= 0.7')
@@ -459,6 +467,7 @@ summary({
     'Snap': enable_snap,
     'Malcontent': enable_malcontent,
     'Whoopsie': enable_whoopsie,
+    'Insights': enable_insights,
   },
   section: 'Optional Dependencies',
   bool_yn: true,
diff --git a/meson_options.txt b/meson_options.txt
index 4e1f5d9..e88d3b3 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -10,3 +10,4 @@ option('malcontent', type: 'boolean', value: false, description: 'build with mal
 option('distributor_logo', type: 'string', description: 'absolute path to distributor logo for the About panel')
 option('dark_mode_distributor_logo', type: 'string', description: 'absolute path to distributor logo dark mode variant')
 option('whoopsie', type: 'boolean', value: false, description: 'build with Whoopsie support')
+option('insights', type: 'boolean', value: false, description: 'build with Ubuntu Insights support')
diff --git a/panels/privacy/cc-privacy-panel.blp b/panels/privacy/cc-privacy-panel.blp
index 4ef1b86..7e81a7e 100644
--- a/panels/privacy/cc-privacy-panel.blp
+++ b/panels/privacy/cc-privacy-panel.blp
@@ -43,13 +43,13 @@ template $CcPrivacyPanel: $CcPanel {
           }
 
           $CcListRow {
-            title: _("_Diagnostics");
-            subtitle: _("Automatic problem reporting");
-            icon-name: "org.gnome.Settings-device-diagnostics-symbolic";
+            title: _("Tele_metry");
+            subtitle: _("Control error and system reporting");
+            icon-name: "diagnostics-symbolic";
             show-arrow: true;
             action-name: "navigation.push";
-            action-target: "'diagnostics'";
-            visible: bind diagnostics_page.visible;
+            action-target: "'telemetry'";
+            visible: bind telemetry_page.visible;
           }
 
           $CcListRow {
@@ -108,7 +108,7 @@ template $CcPrivacyPanel: $CcPanel {
   $CcUsagePage {}
 
   [subpage]
-  $CcDiagnosticsPage diagnostics_page {}
+  $CcTelemetryPage telemetry_page {}
 
   [subpage]
   $CcConnectivityPage {}
diff --git a/panels/privacy/cc-privacy-panel.c b/panels/privacy/cc-privacy-panel.c
index f61d0d8..75e893d 100644
--- a/panels/privacy/cc-privacy-panel.c
+++ b/panels/privacy/cc-privacy-panel.c
@@ -27,7 +27,7 @@
 #endif
 #include "camera/cc-camera-page.h"
 #include "connectivity/cc-connectivity-page.h"
-#include "diagnostics/cc-diagnostics-page.h"
+#include "telemetry/cc-telemetry-page.h"
 #include "firmware-security/cc-firmware-security-page.h"
 #include "cc-list-row.h"
 #include "location/cc-location-page.h"
@@ -74,7 +74,7 @@ cc_privacy_panel_class_init (CcPrivacyPanelClass *klass)
 
   g_type_ensure (CC_TYPE_CAMERA_PAGE);
   g_type_ensure (CC_TYPE_CONNECTIVITY_PAGE);
-  g_type_ensure (CC_TYPE_DIAGNOSTICS_PAGE);
+  g_type_ensure (CC_TYPE_TELEMETRY_PAGE);
   g_type_ensure (CC_TYPE_FIRMWARE_SECURITY_PAGE);
   g_type_ensure (CC_TYPE_LOCATION_PAGE);
   g_type_ensure (CC_TYPE_SCREEN_PAGE);
diff --git a/panels/privacy/gnome-privacy-panel.desktop.in b/panels/privacy/gnome-privacy-panel.desktop.in
index 3704b2b..e44caf3 100644
--- a/panels/privacy/gnome-privacy-panel.desktop.in
+++ b/panels/privacy/gnome-privacy-panel.desktop.in
@@ -11,4 +11,4 @@ StartupNotify=true
 Categories=GNOME;GTK;Settings;DesktopSettings;X-GNOME-Settings-Panel;X-GNOME-AccountSettings;
 OnlyShowIn=GNOME;
 # Translators: Search terms to find the Privacy panel. Do NOT translate or localize the semicolons! The list MUST also end with a semicolon!
-Keywords=Privacy;Screen;Lock;Private;Usage;Recent;History;Files;Temporary;Tmp;Trash;Purge;Retain;Diagnostics;Crash;Location;Gps;Camera;Photos;Video;Webcam;Recording;Security;Firmware;Thunderbolt;Network;error;bug;whoopsie;launchpad;issue;
+Keywords=Privacy;Screen;Lock;Private;Usage;Recent;History;Files;Telemetry;Temporary;Tmp;Trash;Purge;Retain;Diagnostics;Crash;Location;Gps;Camera;Photos;Video;Webcam;Recording;Security;Firmware;Thunderbolt;Network;error;bug;whoopsie;launchpad;issue;
diff --git a/panels/privacy/meson.build b/panels/privacy/meson.build
index 03078c7..5bdea29 100644
--- a/panels/privacy/meson.build
+++ b/panels/privacy/meson.build
@@ -15,7 +15,7 @@ blueprints = []
 subdir('bolt')
 subdir('camera')
 subdir('connectivity')
-subdir('diagnostics')
+subdir('telemetry')
 subdir('firmware-security')
 subdir('icons')
 subdir('location')
@@ -40,7 +40,7 @@ blueprints += custom_target(
 sources = files(
   'camera/cc-camera-page.c',
   'connectivity/cc-connectivity-page.c',
-  'diagnostics/cc-diagnostics-page.c',
+  'telemetry/cc-telemetry-page.c',
   'firmware-security/cc-firmware-security-boot-dialog.c',
   'firmware-security/cc-firmware-security-dialog.c',
   'firmware-security/cc-firmware-security-help-dialog.c',
@@ -52,6 +52,10 @@ sources = files(
   'usage/cc-usage-page.c',
 )
 
+if enable_insights
+sources += files(  'telemetry/cc-telemetry-insights-example.c')
+endif
+
 if host_is_linux_not_s390
   sources += files(
     'bolt/bolt-client.c',
@@ -78,9 +82,22 @@ if host_is_linux_not_s390
   )
 endif
 
+resource_conf = configuration_data()
+telemetry_ui = ''
+if enable_insights
+  telemetry_ui = '    <file preprocess="xml-stripblanks">telemetry/cc-telemetry-insights-example.ui</file>'
+endif
+resource_conf.set('telemetry_insights_example_ui', telemetry_ui)
+
+resource_file = configure_file(
+  input: cappletname + '.gresource.xml.in',
+  output: cappletname + '.gresource.xml',
+  configuration: resource_conf
+)
+
 sources += gnome.compile_resources(
   'cc-' + cappletname + '-resources',
-  cappletname + '.gresource.xml',
+  resource_file,
   c_name: 'cc_' + cappletname,
   export: true,
   dependencies: blueprints,
@@ -107,6 +124,10 @@ if enable_whoopsie
   deps += whoopsie_dep
 endif
 
+if enable_insights
+  deps += insights_dep
+endif
+
 panels_libs += static_library(
   cappletname,
   sources: sources,
diff --git a/panels/privacy/privacy.gresource.xml b/panels/privacy/privacy.gresource.xml.in
similarity index 89%
copy from panels/privacy/privacy.gresource.xml
copy to panels/privacy/privacy.gresource.xml.in
index 3d9612c..2d44ce9 100644
--- a/panels/privacy/privacy.gresource.xml
+++ b/panels/privacy/privacy.gresource.xml.in
@@ -6,7 +6,8 @@
     <file preprocess="xml-stripblanks">bolt/cc-bolt-page.ui</file>
     <file preprocess="xml-stripblanks">camera/cc-camera-page.ui</file>
     <file preprocess="xml-stripblanks">connectivity/cc-connectivity-page.ui</file>
-    <file preprocess="xml-stripblanks">diagnostics/cc-diagnostics-page.ui</file>
+    <file preprocess="xml-stripblanks">telemetry/cc-telemetry-page.ui</file>
+@telemetry_insights_example_ui@
     <file preprocess="xml-stripblanks">firmware-security/cc-firmware-security-boot-dialog.ui</file>
     <file preprocess="xml-stripblanks">firmware-security/cc-firmware-security-dialog.ui</file>
     <file preprocess="xml-stripblanks">firmware-security/cc-firmware-security-help-dialog.ui</file>
@@ -16,6 +17,7 @@
     <file preprocess="xml-stripblanks">screen/cc-screen-page.ui</file>
     <file preprocess="xml-stripblanks">usage/cc-usage-page.ui</file>
 
+    <file>telemetry/telemetry.css</file>
     <file>firmware-security/security-level.css</file>
   </gresource>
 </gresources>
diff --git a/panels/privacy/telemetry/cc-telemetry-insights-example.blp b/panels/privacy/telemetry/cc-telemetry-insights-example.blp
new file mode 100644
index 0000000..3f5ce2f
--- /dev/null
+++ b/panels/privacy/telemetry/cc-telemetry-insights-example.blp
@@ -0,0 +1,47 @@
+using Gtk 4.0;
+using Adw 1;
+
+template $CcTelemetryInsightsExample: Adw.Window {
+  title: _("System Report");
+  modal: true;
+  width-request: 535;
+  height-request: 600;
+
+  content: Adw.ToolbarView {
+    [top]
+    Adw.HeaderBar {}
+
+    content: ScrolledWindow {
+      hscrollbar-policy: never;
+
+      child: Box {
+        orientation: vertical;
+        spacing: 12;
+        margin-top: 29;
+        margin-bottom: 29;
+        margin-start: 24;
+        margin-end: 24;
+
+        Label subtitle_label {
+          wrap: true;
+          wrap-mode: word_char;
+          valign: center;
+          halign: start;
+        }
+
+        TextView report_view {
+          editable: false;
+          cursor-visible: false;
+          monospace: true;
+          wrap-mode: word_char;
+          left-margin: 12;
+          right-margin: 12;
+          top-margin: 12;
+          bottom-margin: 12;
+
+          buffer: Gtk.TextBuffer {};
+        }
+      };
+    };
+  };
+}
diff --git a/panels/privacy/telemetry/cc-telemetry-insights-example.c b/panels/privacy/telemetry/cc-telemetry-insights-example.c
new file mode 100644
index 0000000..c3c36b0
--- /dev/null
+++ b/panels/privacy/telemetry/cc-telemetry-insights-example.c
@@ -0,0 +1,102 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright (C) 2026 Canonical Ltd.
+ *
+ * Written by: Kat Kuo <kat.kuo@canonical.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <config.h>
+
+#include <glib/gi18n.h>
+
+#include "cc-telemetry-insights-example.h"
+#include "glib.h"
+#include "gtk/gtk.h"
+#include "gtk/gtkshortcut.h"
+
+#define DOCUMENTATION_URL "https://github.com/ubuntu/ubuntu-insights#example-insights-reports"
+#define INSIGHTS_CACHE_SUBDIR ".cache/ubuntu-insights"
+
+struct _CcTelemetryInsightsExample
+{
+  AdwWindow parent_instance;
+
+  GtkWidget   *subtitle_label;
+  GtkTextView *report_view;
+};
+
+G_DEFINE_TYPE (CcTelemetryInsightsExample, cc_telemetry_insights_example, ADW_TYPE_WINDOW)
+
+static void
+cc_telemetry_insights_example_class_init (CcTelemetryInsightsExampleClass *klass)
+{
+  GtkWidgetClass *widget_class = GTK_WIDGET_CLASS (klass);
+
+  gtk_widget_class_add_binding_action (widget_class, GDK_KEY_Escape, 0, "window.close", NULL);
+
+  gtk_widget_class_set_template_from_resource (widget_class,
+                                               "/org/gnome/control-center/privacy/telemetry/cc-telemetry-insights-example.ui");
+
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryInsightsExample, subtitle_label);
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryInsightsExample, report_view);
+}
+
+static void
+cc_telemetry_insights_example_init (CcTelemetryInsightsExample *self)
+{
+  g_autofree gchar *path = NULL;
+  g_autofree gchar *url = NULL;
+  g_autofree gchar *subtitle = NULL;
+  g_autofree gchar *cache_path = NULL;
+  g_autofree gchar *cache_uri = NULL;
+  g_autofree gchar *display_path = NULL;
+
+  gtk_widget_init_template (GTK_WIDGET (self));
+
+  cache_path = g_build_filename (g_get_home_dir (), INSIGHTS_CACHE_SUBDIR, NULL);
+  display_path = g_strdup_printf ("~/%s", INSIGHTS_CACHE_SUBDIR);
+
+  if (g_file_test (cache_path, G_FILE_TEST_IS_DIR))
+    {
+      cache_uri = g_filename_to_uri (cache_path, NULL, NULL);
+      path = g_strdup_printf ("<a href=\"%s\"><span face=\"monospace\" allow_breaks=\"false\">%s</span></a>",
+                              cache_uri,
+                              display_path);
+    }
+  else
+    {
+      path = g_strdup_printf ("<span face=\"monospace\" allow_breaks=\"false\">%s</span>",
+                              display_path);
+    }
+
+  /* translators: The '%s' is a file path to the local Ubuntu Insights cache directory */
+  subtitle = g_strdup_printf (_ ("This is a system report example with real data from this computer. "
+                                 "System reports are temporarily stored in %s before being sent to Canonical."),
+                              path);
+  gtk_label_set_markup (GTK_LABEL (self->subtitle_label), subtitle);
+}
+
+GtkWidget *
+cc_telemetry_insights_example_new (const gchar *report)
+{
+  CcTelemetryInsightsExample *self;
+  self = g_object_new (CC_TYPE_TELEMETRY_INSIGHTS_EXAMPLE, NULL);
+
+  gtk_text_buffer_set_text (gtk_text_view_get_buffer (self->report_view),
+                            report,
+                            -1);
+  return GTK_WIDGET (self);
+}
diff --git a/panels/mouse/cc-mouse-test.h b/panels/privacy/telemetry/cc-telemetry-insights-example.h
similarity index 62%
copy from panels/mouse/cc-mouse-test.h
copy to panels/privacy/telemetry/cc-telemetry-insights-example.h
index ccd7515..bb5f6cc 100644
--- a/panels/mouse/cc-mouse-test.h
+++ b/panels/privacy/telemetry/cc-telemetry-insights-example.h
@@ -1,8 +1,8 @@
-/* -*- mode: c; style: linux -*-
- * 
- * Copyright (C) 2012 Red Hat, Inc.
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
  *
- * Written by: Ondrej Holy <oholy@redhat.com>
+ * Copyright (C) 2026 Canonical Ltd.
+ *
+ * Written by: Kat Kuo <kat.kuo@canonical.com>
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -24,9 +24,9 @@
 
 G_BEGIN_DECLS
 
-#define CC_TYPE_MOUSE_TEST (cc_mouse_test_get_type ())
-G_DECLARE_FINAL_TYPE (CcMouseTest, cc_mouse_test, CC, MOUSE_TEST, AdwWindow)
+#define CC_TYPE_TELEMETRY_INSIGHTS_EXAMPLE (cc_telemetry_insights_example_get_type ())
+G_DECLARE_FINAL_TYPE (CcTelemetryInsightsExample, cc_telemetry_insights_example, CC, TELEMETRY_INSIGHTS_EXAMPLE, AdwWindow)
 
-GtkWidget *cc_mouse_test_new (void);
+GtkWidget *cc_telemetry_insights_example_new (const gchar *report);
 
 G_END_DECLS
diff --git a/panels/privacy/telemetry/cc-telemetry-page.blp b/panels/privacy/telemetry/cc-telemetry-page.blp
new file mode 100644
index 0000000..dca5f7b
--- /dev/null
+++ b/panels/privacy/telemetry/cc-telemetry-page.blp
@@ -0,0 +1,90 @@
+using Gtk 4.0;
+using Adw 1;
+
+template $CcTelemetryPage: Adw.NavigationPage {
+  title: _("Telemetry");
+  tag: "telemetry";
+  visible: false;
+
+  child: Adw.ToolbarView {
+    [top]
+    Adw.HeaderBar {}
+
+    content: Adw.PreferencesPage {
+      Adw.PreferencesGroup diagnostics_group {
+        title: _("Error Reporting");
+
+        Adw.SwitchRow abrt_row {
+          title: _("_Automatic Error Reporting");
+          use-underline: true;
+        }
+
+        Adw.ComboRow whoopsie_combo_row {
+          title: _("Send error reports to Canonical");
+          visible: false;
+          use-underline: true;
+          notify::selected-item => $whoopsie_combo_row_changed_cb(template);
+        }
+
+        LinkButton diagnostics_learn_more_link_button {
+          label: _("Learn more about error reporting");
+          visible: bind $string_is_not_empty(diagnostics_learn_more_link_button.uri) as <bool>;
+          margin-top: 8;
+          halign: start;
+
+          styles [
+            "dim",
+            "body",
+            "link-label",
+          ]
+        }
+      }
+
+      Adw.PreferencesGroup insights_group {
+        title: _("System Reporting");
+        visible: false;
+
+        Adw.SwitchRow insights_linux_row {
+          title: _("Send a system report every month");
+          notify::active => $insights_switch_row_changed_cb();
+        }
+
+        Adw.SwitchRow insights_upgrade_row {
+          title: _("Send a system report when upgrading to a new Ubuntu version");
+          notify::active => $insights_switch_row_changed_cb();
+        }
+      }
+
+      Adw.PreferencesGroup insights_buttons_group {
+        separate-rows: true;
+        visible: bind insights_group.visible;
+
+        Adw.ButtonRow insights_example_report_button_row {
+          title: _("_Show Report");
+          use-underline: true;
+          end-icon-name: "go-next-symbolic";
+          activated => $insights_example_report_button_row_activated_cb(template);
+        }
+
+        Adw.ButtonRow insights_cancel_report_button_row {
+          title: _("Cancel Report Generation");
+          visible: false;
+          activated => $insights_cancel_report_button_row_activated_cb(template);
+        }
+
+        LinkButton insights_learn_more_link_button {
+          label: _("Learn more about system reporting");
+          uri: "https://canonical.com/legal/systems-information-notice";
+          margin-top: 8;
+          halign: start;
+
+          styles [
+            "dim",
+            "body",
+            "link-label",
+          ]
+        }
+      }
+    };
+  };
+}
diff --git a/panels/privacy/telemetry/cc-telemetry-page.c b/panels/privacy/telemetry/cc-telemetry-page.c
new file mode 100644
index 0000000..0085ad3
--- /dev/null
+++ b/panels/privacy/telemetry/cc-telemetry-page.c
@@ -0,0 +1,629 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright (C) 2026 Canonical Ltd.
+ * Copyright (C) 2018 Red Hat, Inc
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author(s):
+ *   Kat Kuo <kat.kuo@canonical.com>
+ *   Matthias Clasen <mclasen@redhat.com>
+ */
+
+#include <config.h>
+
+#include "cc-telemetry-page.h"
+#include "cc-util.h"
+#include "gtk/gtk.h"
+#include "shell/cc-application.h"
+
+#include <gio/gdesktopappinfo.h>
+#include <glib/gi18n.h>
+#include <stdio.h>
+
+#ifdef HAVE_WHOOPSIE
+#include <whoopsie-preferences/libwhoopsie-preferences.h>
+#endif
+
+#ifdef HAVE_INSIGHTS
+#include "cc-telemetry-insights-example.h"
+#include <insights/insights.h>
+#include <insights/types.h>
+#endif
+
+struct _CcTelemetryPage
+{
+  AdwNavigationPage parent_instance;
+
+  AdwPreferencesGroup *diagnostics_group;
+  AdwSwitchRow        *abrt_row;
+
+#ifdef HAVE_WHOOPSIE
+  AdwComboRow         *whoopsie_combo_row;
+  WhoopsiePreferences *whoopsie;
+#endif
+  GtkLinkButton *diagnostics_learn_more_link_button;
+
+  AdwPreferencesGroup *insights_group;
+  AdwPreferencesGroup *insights_buttons_group;
+
+#ifdef HAVE_INSIGHTS
+  AdwSwitchRow *insights_linux_row;
+  AdwSwitchRow *insights_upgrade_row;
+
+  AdwButtonRow *insights_example_report_button_row;
+  AdwButtonRow *insights_cancel_report_button_row;
+#endif
+
+  GCancellable *cancellable;
+
+#ifdef HAVE_INSIGHTS
+  GCancellable *insights_cancellable;
+#endif
+
+  GSettings *privacy_settings;
+};
+
+G_DEFINE_TYPE (CcTelemetryPage, cc_telemetry_page, ADW_TYPE_NAVIGATION_PAGE)
+
+/* Static init function */
+
+static gboolean
+string_is_not_empty (CcTelemetryPage *self,
+                     const char      *string)
+{
+  return string && string[0];
+}
+
+static void
+abrt_appeared_cb (GDBusConnection *connection,
+                  const gchar     *name,
+                  const gchar     *name_owner,
+                  gpointer         user_data)
+{
+  g_autofree gchar *whoopsie_url = NULL;
+  CcTelemetryPage  *self = CC_TELEMETRY_PAGE (user_data);
+
+  g_debug ("ABRT appeared");
+  gtk_widget_set_visible (GTK_WIDGET (self), TRUE);
+  gtk_widget_set_visible (GTK_WIDGET (self->diagnostics_group), TRUE);
+}
+
+static void
+abrt_vanished_cb (GDBusConnection *connection,
+                  const gchar     *name,
+                  gpointer         user_data)
+{
+  g_debug ("ABRT vanished");
+#ifndef HAVE_WHOOPSIE
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (user_data);
+  gtk_widget_set_visible (GTK_WIDGET (self->diagnostics_group), FALSE);
+
+#ifndef HAVE_INSIGHTS
+  gtk_widget_set_visible (GTK_WIDGET (self), FALSE);
+#endif
+#endif
+}
+
+#ifdef HAVE_WHOOPSIE
+typedef enum
+{
+  WHOOPSIE_BUTTON_SETTING_NEVER,
+  WHOOPSIE_BUTTON_SETTING_AUTO,
+  WHOOPSIE_BUTTON_SETTING_MANUAL,
+} WhoopsieButtonSettingType;
+
+static void
+whoopsie_set_report_crashes_done (GObject      *source_object,
+                                  GAsyncResult *res,
+                                  gpointer      user_data)
+{
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (user_data);
+  g_autoptr (GError) error = NULL;
+
+  if (!whoopsie_preferences_call_set_report_crashes_finish (self->whoopsie, res, &error))
+    {
+      if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        return;
+
+      g_warning ("Failed to toggle crash reporting: %s", error->message);
+    }
+}
+
+static void
+whoopsie_set_report_crashes_mode_done (GObject      *source_object,
+                                       GAsyncResult *res,
+                                       gpointer      user_data)
+{
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (user_data);
+  g_autoptr (GError) error = NULL;
+
+  if (!whoopsie_preferences_call_set_automatically_report_crashes_finish (self->whoopsie, res, &error))
+    {
+      if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        return;
+
+      g_warning ("Failed to toggle crash reporting mode: %s", error->message);
+    }
+}
+
+static void
+whoopsie_combo_row_changed_cb (CcTelemetryPage *self)
+{
+  g_autoptr (GObject) item = NULL;
+  GListModel *model;
+  gint        selected_index;
+  gint        value;
+
+  model = adw_combo_row_get_model (self->whoopsie_combo_row);
+  selected_index = adw_combo_row_get_selected (self->whoopsie_combo_row);
+  if (selected_index == -1)
+    return;
+
+  item = g_list_model_get_item (model, selected_index);
+  value = GPOINTER_TO_UINT (g_object_get_data (G_OBJECT (item), "value"));
+
+  whoopsie_preferences_call_set_report_crashes (self->whoopsie,
+                                                value != WHOOPSIE_BUTTON_SETTING_NEVER,
+                                                self->cancellable,
+                                                whoopsie_set_report_crashes_done,
+                                                self);
+
+  whoopsie_preferences_call_set_automatically_report_crashes (self->whoopsie,
+                                                              value == WHOOPSIE_BUTTON_SETTING_AUTO,
+                                                              self->cancellable,
+                                                              whoopsie_set_report_crashes_mode_done,
+                                                              self);
+}
+
+static void
+set_value_for_combo_row (AdwComboRow *combo_row, gint value)
+{
+  g_autoptr (GObject) new_item = NULL;
+  GListModel *model;
+  guint       i;
+
+  /* try to make the UI match the setting */
+  model = adw_combo_row_get_model (combo_row);
+  for (i = 0; i < g_list_model_get_n_items (model); i++)
+    {
+      g_autoptr (GObject) item = g_list_model_get_item (model, i);
+      gint value_tmp = GPOINTER_TO_UINT (g_object_get_data (item, "value"));
+      if (value_tmp == value)
+        {
+          adw_combo_row_set_selected (combo_row, i);
+          return;
+        }
+    }
+}
+
+static void
+whoopsie_properties_changed (CcTelemetryPage *self)
+{
+  g_autofree char *name_owner = g_dbus_proxy_get_name_owner (G_DBUS_PROXY (self->whoopsie));
+
+  /* Bail out if dbus service has gone away */
+  if (G_UNLIKELY (!name_owner))
+    return;
+
+  WhoopsieButtonSettingType value = WHOOPSIE_BUTTON_SETTING_NEVER;
+
+  if (whoopsie_preferences_get_automatically_report_crashes (self->whoopsie))
+    value = WHOOPSIE_BUTTON_SETTING_AUTO;
+  else if (whoopsie_preferences_get_report_crashes (self->whoopsie))
+    value = WHOOPSIE_BUTTON_SETTING_MANUAL;
+
+  g_signal_handlers_block_by_func (self->whoopsie_combo_row, whoopsie_combo_row_changed_cb, self);
+  set_value_for_combo_row (self->whoopsie_combo_row, value);
+  g_signal_handlers_unblock_by_func (self->whoopsie_combo_row, whoopsie_combo_row_changed_cb, self);
+}
+
+static void
+whoopsie_bus_owner_changed (GDBusProxy *proxy)
+{
+  g_autofree char *name_owner = g_dbus_proxy_get_name_owner (proxy);
+
+  if (!name_owner)
+    {
+      /* Trigger the service to start again, as it has just been stopped by a timeout
+       * don't bother on waiting for the call results.
+       */
+      whoopsie_preferences_call_get_identifier (WHOOPSIE_PREFERENCES (proxy), NULL, NULL, NULL);
+    }
+}
+
+static void
+populate_whoopsie_button_row (AdwComboRow *combo_row)
+{
+  g_autoptr (GtkStringList) string_list = NULL;
+  struct
+  {
+    char                     *name;
+    WhoopsieButtonSettingType value;
+  } actions[] = {
+    { NC_ ("Whoopsie error reporting", "Never"), WHOOPSIE_BUTTON_SETTING_NEVER },
+    { NC_ ("Whoopsie error reporting", "Automatic"), WHOOPSIE_BUTTON_SETTING_AUTO },
+    { NC_ ("Whoopsie error reporting", "Manual"), WHOOPSIE_BUTTON_SETTING_MANUAL },
+  };
+  guint item_index = 0;
+  guint i;
+
+  string_list = gtk_string_list_new (NULL);
+  for (i = 0; i < G_N_ELEMENTS (actions); i++)
+    {
+      g_autoptr (GObject) item = NULL;
+
+      gtk_string_list_append (string_list, _ (actions[i].name));
+
+      item = g_list_model_get_item (G_LIST_MODEL (string_list), item_index++);
+      g_object_set_data (item, "value", GUINT_TO_POINTER (actions[i].value));
+    }
+
+  adw_combo_row_set_model (combo_row, G_LIST_MODEL (string_list));
+}
+
+static void
+on_new_whoopsie_proxy_cb (GObject      *object,
+                          GAsyncResult *res,
+                          gpointer      data)
+{
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (data);
+  g_autoptr (GError) error = NULL;
+
+  self->whoopsie = whoopsie_preferences_proxy_new_for_bus_finish (res, &error);
+
+  if (error)
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        g_warning ("Failed to get whoopsie proxy: %s", error->message);
+
+      gtk_widget_set_visible (GTK_WIDGET (self->whoopsie_combo_row), FALSE);
+      gtk_widget_set_visible (GTK_WIDGET (self->abrt_row), TRUE);
+      return;
+    }
+
+  g_debug ("Whoopsie available");
+  gtk_widget_set_visible (GTK_WIDGET (self), TRUE);
+  gtk_widget_set_visible (GTK_WIDGET (self->whoopsie_combo_row), TRUE);
+  gtk_widget_set_sensitive (GTK_WIDGET (self->whoopsie_combo_row), TRUE);
+
+  g_signal_handlers_block_by_func (self->whoopsie_combo_row,
+                                   whoopsie_combo_row_changed_cb, self);
+  populate_whoopsie_button_row (self->whoopsie_combo_row);
+  g_signal_handlers_unblock_by_func (self->whoopsie_combo_row,
+                                     whoopsie_combo_row_changed_cb, self);
+
+  g_signal_connect_object (self->whoopsie, "g-properties-changed",
+                           G_CALLBACK (whoopsie_properties_changed),
+                           self, G_CONNECT_SWAPPED);
+
+  g_signal_connect_object (self->whoopsie, "notify::g-name-owner",
+                           G_CALLBACK (whoopsie_bus_owner_changed),
+                           self, 0);
+
+  whoopsie_properties_changed (self);
+}
+#else
+static void
+whoopsie_combo_row_changed_cb (CcTelemetryPage *self)
+{
+}
+#endif
+
+#ifdef HAVE_INSIGHTS
+static void update_insights_row_state (CcTelemetryPage *self, AdwSwitchRow *row);
+
+static void
+insights_switch_row_changed_cb (GObject    *object,
+                                GParamSpec *pspec,
+                                gpointer    user_data)
+{
+  AdwSwitchRow    *row = ADW_SWITCH_ROW (object);
+  gboolean         active;
+  const gchar     *source;
+  g_autofree char *error_msg = NULL;
+
+  active = adw_switch_row_get_active (row);
+  source = g_object_get_data (G_OBJECT (row), "insights-source");
+
+  if (!source)
+    return; /* Shouldn't happen */
+
+  error_msg = insights_set_consent_state (NULL,
+                                          source,
+                                          active ? INSIGHTS_CONSENT_TRUE : INSIGHTS_CONSENT_FALSE);
+  if (error_msg != NULL)
+    {
+      g_warning ("Failed to set insights consent state for source '%s': %s", source, error_msg);
+      update_insights_row_state (CC_TELEMETRY_PAGE (user_data), row);
+    }
+}
+
+static void
+update_insights_row_state (CcTelemetryPage *self,
+                           AdwSwitchRow    *row)
+{
+  const char            *source;
+  insights_consent_state state;
+
+  source = g_object_get_data (G_OBJECT (row), "insights-source");
+  if (!source)
+    return; /* Shouldn't happen */
+
+  state = insights_get_consent_state (NULL, source);
+  g_signal_handlers_block_by_func (row, insights_switch_row_changed_cb, self);
+  adw_switch_row_set_active (row, state == INSIGHTS_CONSENT_TRUE);
+  g_signal_handlers_unblock_by_func (row, insights_switch_row_changed_cb, self);
+}
+
+static void
+sync_insights_switches (CcTelemetryPage *self)
+{
+  update_insights_row_state (self, self->insights_linux_row);
+  update_insights_row_state (self, self->insights_upgrade_row);
+}
+
+static void
+on_generate_report_thread (GTask        *task,
+                           gpointer      source_object,
+                           gpointer      task_data,
+                           GCancellable *cancellable)
+{
+  g_autoptr (GError) error = NULL;
+  g_autofree char       *error_msg = NULL;
+  g_autofree char       *report = NULL;
+  insights_compile_flags flags = { 0 };
+
+  error_msg = insights_compile (NULL, &flags, &report);
+  if (error_msg != NULL)
+    {
+      g_task_return_new_error (task, G_IO_ERROR, G_IO_ERROR_FAILED, "%s", error_msg);
+      return;
+    }
+
+  g_task_return_pointer (task, g_steal_pointer (&report), g_free);
+}
+
+static void
+generate_report_cb (GObject      *source_object,
+                    GAsyncResult *res,
+                    gpointer      user_data)
+{
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (user_data);
+  g_autofree char *report = NULL;
+  g_autoptr (GError) error = NULL;
+  GtkWidget *window;
+
+  gtk_widget_set_visible (GTK_WIDGET (self->insights_example_report_button_row), TRUE);
+  gtk_widget_set_visible (GTK_WIDGET (self->insights_cancel_report_button_row), FALSE);
+
+  report = g_task_propagate_pointer (G_TASK (res), &error);
+
+  if (error)
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        g_warning ("Failed to generate example report: %s", error->message);
+      return;
+    }
+
+  window = cc_telemetry_insights_example_new (report);
+  gtk_window_set_transient_for (GTK_WINDOW (window),
+                                GTK_WINDOW (gtk_widget_get_native (GTK_WIDGET (self))));
+  gtk_window_present (GTK_WINDOW (window));
+}
+
+static void
+insights_example_report_button_row_activated_cb (CcTelemetryPage *self)
+{
+  g_autoptr (GTask) task;
+
+  gtk_widget_set_visible (GTK_WIDGET (self->insights_example_report_button_row), FALSE);
+  gtk_widget_set_visible (GTK_WIDGET (self->insights_cancel_report_button_row), TRUE);
+
+  task = g_task_new (self, self->insights_cancellable, generate_report_cb, self);
+
+  g_task_set_return_on_cancel (task, TRUE);
+  g_task_run_in_thread (task, on_generate_report_thread);
+}
+
+static void
+insights_cancel_report_button_row_activated_cb (CcTelemetryPage *self)
+{
+  g_debug ("User cancelled insights report generation");
+  gtk_widget_set_visible (GTK_WIDGET (self->insights_cancel_report_button_row), FALSE);
+  gtk_widget_set_visible (GTK_WIDGET (self->insights_example_report_button_row), TRUE);
+
+  g_cancellable_cancel (self->insights_cancellable);
+  g_clear_object (&self->insights_cancellable);
+  self->insights_cancellable = g_cancellable_new ();
+}
+
+#else
+static void
+insights_switch_row_changed_cb (GObject    *object,
+                                GParamSpec *pspec,
+                                gpointer    user_data)
+{
+}
+
+static void
+insights_example_report_button_row_activated_cb (CcTelemetryPage *self)
+{
+}
+
+static void
+insights_cancel_report_button_row_activated_cb (CcTelemetryPage *self)
+{
+}
+
+#endif
+
+static void
+cc_telemetry_page_finalize (GObject *object)
+{
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (object);
+
+  g_clear_object (&self->privacy_settings);
+
+  g_cancellable_cancel (self->cancellable);
+  g_clear_object (&self->cancellable);
+
+#ifdef HAVE_WHOOPSIE
+  g_clear_object (&self->whoopsie);
+#endif
+
+#ifdef HAVE_INSIGHTS
+  g_cancellable_cancel (self->insights_cancellable);
+  g_clear_object (&self->insights_cancellable);
+#endif
+
+  G_OBJECT_CLASS (cc_telemetry_page_parent_class)->finalize (object);
+}
+
+static void
+cc_telemetry_page_class_init (CcTelemetryPageClass *klass)
+{
+  GObjectClass   *oclass = G_OBJECT_CLASS (klass);
+  GtkWidgetClass *widget_class = GTK_WIDGET_CLASS (klass);
+
+  oclass->finalize = cc_telemetry_page_finalize;
+
+  gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/control-center/privacy/telemetry/cc-telemetry-page.ui");
+
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, diagnostics_group);
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, abrt_row);
+
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, insights_group);
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, insights_buttons_group);
+
+#ifdef HAVE_WHOOPSIE
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, whoopsie_combo_row);
+#endif
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, diagnostics_learn_more_link_button);
+  gtk_widget_class_bind_template_callback (widget_class, whoopsie_combo_row_changed_cb);
+
+#ifdef HAVE_INSIGHTS
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, insights_linux_row);
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, insights_upgrade_row);
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, insights_example_report_button_row);
+  gtk_widget_class_bind_template_child (widget_class, CcTelemetryPage, insights_cancel_report_button_row);
+#endif
+  gtk_widget_class_bind_template_callback (widget_class, insights_switch_row_changed_cb);
+  gtk_widget_class_bind_template_callback (widget_class, insights_example_report_button_row_activated_cb);
+  gtk_widget_class_bind_template_callback (widget_class, insights_cancel_report_button_row_activated_cb);
+
+  gtk_widget_class_bind_template_callback (widget_class, string_is_not_empty);
+}
+
+#ifdef HAVE_INSIGHTS
+static void
+on_unmap_cb (GtkWidget *widget,
+             gpointer   user_data)
+{
+  CcTelemetryPage *self = CC_TELEMETRY_PAGE (widget);
+
+  if (self->insights_cancellable)
+    {
+      g_cancellable_cancel (self->insights_cancellable);
+      g_object_unref (self->insights_cancellable);
+      self->insights_cancellable = g_cancellable_new ();
+    }
+}
+#endif
+
+static void
+cc_telemetry_page_init (CcTelemetryPage *self)
+{
+  g_autofree gchar *os_name = NULL;
+  g_autofree gchar *whoopsie_url = NULL;
+  g_autofree gchar *whoopsie_msg = NULL;
+  g_autofree gchar *insights_msg = NULL;
+  g_autoptr (GtkCssProvider) provider = NULL;
+
+  gtk_widget_init_template (GTK_WIDGET (self));
+
+#ifdef HAVE_INSIGHTS
+  g_object_set_data (G_OBJECT (self->insights_linux_row), "insights-source", "linux");
+  g_object_set_data (G_OBJECT (self->insights_upgrade_row), "insights-source", "ubuntu_release_upgrader");
+#endif
+
+  g_bus_watch_name (G_BUS_TYPE_SYSTEM,
+                    "org.freedesktop.problems.daemon",
+                    G_BUS_NAME_WATCHER_FLAGS_NONE,
+                    abrt_appeared_cb,
+                    abrt_vanished_cb,
+                    self,
+                    NULL);
+
+  self->privacy_settings = g_settings_new ("org.gnome.desktop.privacy");
+
+  g_settings_bind (self->privacy_settings, "report-technical-problems",
+                   self->abrt_row, "active",
+                   G_SETTINGS_BIND_DEFAULT);
+
+#if defined(HAVE_WHOOPSIE) || defined(HAVE_INSIGHTS)
+  self->cancellable = g_cancellable_new ();
+#endif
+
+#ifdef HAVE_WHOOPSIE
+  /* check for whoopsie */
+  whoopsie_preferences_proxy_new_for_bus (G_BUS_TYPE_SYSTEM,
+                                          G_DBUS_PROXY_FLAGS_NONE,
+                                          "com.ubuntu.WhoopsiePreferences",
+                                          "/com/ubuntu/WhoopsiePreferences",
+                                          self->cancellable,
+                                          on_new_whoopsie_proxy_cb,
+                                          self);
+
+  gtk_widget_set_visible (GTK_WIDGET (self), TRUE);
+  gtk_widget_set_visible (GTK_WIDGET (self->whoopsie_combo_row), TRUE);
+  gtk_widget_set_visible (GTK_WIDGET (self->abrt_row), FALSE);
+  gtk_widget_set_sensitive (GTK_WIDGET (self->whoopsie_combo_row), FALSE);
+#endif
+
+#ifdef HAVE_INSIGHTS
+  self->insights_cancellable = g_cancellable_new ();
+  gtk_widget_set_visible (GTK_WIDGET (self), TRUE);
+  gtk_widget_set_visible (GTK_WIDGET (self->insights_group), TRUE);
+
+  g_signal_connect_swapped (self, "map", G_CALLBACK (sync_insights_switches), self);
+  g_signal_connect (self, "unmap", G_CALLBACK (on_unmap_cb), NULL);
+#endif
+
+  os_name = g_get_os_info (G_OS_INFO_KEY_NAME);
+  if (!os_name)
+    os_name = g_strdup ("GNOME");
+  whoopsie_url = g_get_os_info (G_OS_INFO_KEY_PRIVACY_POLICY_URL);
+  if (whoopsie_url)
+    gtk_link_button_set_uri (self->diagnostics_learn_more_link_button, whoopsie_url);
+
+  /* translators: The '%s' is the distributor's name, such as 'Fedora' */
+  whoopsie_msg = g_strdup_printf (_ ("Help us address technical problems by sharing error reports with the "
+                                     "%s team. Reports are sent anonymously and are scrubbed of personal data."),
+                                  os_name);
+  adw_preferences_group_set_description (self->diagnostics_group, whoopsie_msg);
+
+  /* translators: The '%s' is the distributor's name, such as 'Ubuntu' */
+  insights_msg = g_strdup_printf (_ ("Help us improve %s features and compatibility by sharing system reports with"
+                                     " Canonical. Reports are sent anonymously and do not contain any personal data."),
+                                  os_name);
+
+  adw_preferences_group_set_description (self->insights_group, insights_msg);
+
+  provider = gtk_css_provider_new ();
+  gtk_css_provider_load_from_resource (provider, "/org/gnome/control-center/privacy/telemetry/telemetry.css");
+  gtk_style_context_add_provider_for_display (gdk_display_get_default (),
+                                              GTK_STYLE_PROVIDER (provider),
+                                              GTK_STYLE_PROVIDER_PRIORITY_APPLICATION);
+}
diff --git a/panels/privacy/camera/cc-camera-page.h b/panels/privacy/telemetry/cc-telemetry-page.h
similarity index 73%
copy from panels/privacy/camera/cc-camera-page.h
copy to panels/privacy/telemetry/cc-telemetry-page.h
index aed052e..acb5f1a 100644
--- a/panels/privacy/camera/cc-camera-page.h
+++ b/panels/privacy/telemetry/cc-telemetry-page.h
@@ -1,5 +1,6 @@
 /* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
  *
+ * Copyright (C) 2026 Canonical Ltd.
  * Copyright (C) 2018 Red Hat, Inc
  *
  * This program is free software; you can redistribute it and/or modify
@@ -15,7 +16,9 @@
  * You should have received a copy of the GNU General Public License
  * along with this program; if not, see <http://www.gnu.org/licenses/>.
  *
- * Author: Matthias Clasen <mclasen@redhat.com>
+ * Author(s):
+ *   Kat Kuo <kat.kuo@canonical.com>
+ *   Matthias Clasen <mclasen@redhat.com>
  */
 
 #pragma once
@@ -24,7 +27,7 @@
 
 G_BEGIN_DECLS
 
-#define CC_TYPE_CAMERA_PAGE (cc_camera_page_get_type ())
-G_DECLARE_FINAL_TYPE (CcCameraPage, cc_camera_page, CC, CAMERA_PAGE, AdwNavigationPage)
+#define CC_TYPE_TELEMETRY_PAGE (cc_telemetry_page_get_type ())
+G_DECLARE_FINAL_TYPE (CcTelemetryPage, cc_telemetry_page, CC, TELEMETRY_PAGE, AdwNavigationPage)
 
 G_END_DECLS
diff --git a/panels/privacy/diagnostics/meson.build b/panels/privacy/telemetry/meson.build
similarity index 73%
copy from panels/privacy/diagnostics/meson.build
copy to panels/privacy/telemetry/meson.build
index 8fa2539..dffd7e8 100644
--- a/panels/privacy/diagnostics/meson.build
+++ b/panels/privacy/telemetry/meson.build
@@ -1,7 +1,11 @@
 ui_sources = files(
-  'cc-diagnostics-page.blp',
+  'cc-telemetry-page.blp',
 )
 
+if enable_insights
+  ui_sources += files('cc-telemetry-insights-example.blp')
+endif
+
 ui_compiled = []
 foreach ui_source : ui_sources
   ui_compiled += fs.name(fs.replace_suffix(ui_source, '.ui'))
diff --git a/panels/privacy/telemetry/telemetry.css b/panels/privacy/telemetry/telemetry.css
new file mode 100644
index 0000000..281d579
--- /dev/null
+++ b/panels/privacy/telemetry/telemetry.css
@@ -0,0 +1,8 @@
+/* Emulate spacing of a label */
+.link-label {
+  padding: 0;
+  box-shadow: none;
+  border-radius: 0px;
+  min-height: 0px;
+  min-width: 0px;
+}
diff --git a/po/POTFILES.in b/po/POTFILES.in
index 7e660fb..dfb6294 100644
--- a/po/POTFILES.in
+++ b/po/POTFILES.in
@@ -173,6 +173,10 @@ panels/privacy/firmware-security/cc-firmware-security-utils.c
 panels/privacy/gnome-privacy-panel.desktop.in
 panels/privacy/location/cc-location-page.blp
 panels/privacy/screen/cc-screen-page.blp
+panels/privacy/telemetry/cc-telemetry-insights-example.blp
+panels/privacy/telemetry/cc-telemetry-insights-example.c
+panels/privacy/telemetry/cc-telemetry-page.blp
+panels/privacy/telemetry/cc-telemetry-page.c
 panels/privacy/usage/cc-usage-page.blp
 panels/search/cc-search-locations-page.blp
 panels/search/cc-search-locations-page.c
